sfge/src/main/java/com/salesforce/rules/PerformNullCheckOnSoqlVariables.java

```diff
@@ -0,0 +1,142 @@
+    }
+
+    @Override
+    protected String getCategory() {
+        return CATEGORY.PERFORMANCE.name;
+    }
+
+    @Override
+    protected String getUrl() {
+        return URL;
+    }
+
+    public static PerformNullCheckOnSoqlVariables getInstance() {
+        return LazyHolder.INSTANCE;
+    }
+
+    @Override
+    protected boolean isEnabled() {
+        return true;
+    }
+
+    /**
+     * Tests a vertex using a symbol provider to check if it violates this rule.
+     *
+     * @param vertex the vertex to check for this rule
+     * @param symbols a {@link SymbolProvider} that can provide symbols for the given vertex
+     * @return true if, according to the symbol provider, the vertex is equal to null, constrained
+     *     to be null, or has an indeterminate value
+     */
+    @Override
+    public boolean test(BaseSFVertex vertex, SymbolProvider symbols) {
+
+        // We only check for SOQL expressions with variables as grandchildren.
+        // No need to check for MethodCallExpressionVertex because the only relevant method is
+        // Database.queryWithBinds, which already does null checks for the keys and
+        // values in its bind variables.
+        if (!(vertex instanceof VariableExpressionVertex)
+                || !(vertex.getParent().getParent() instanceof SoqlExpressionVertex)) {
+            return false;
+        }
+
+        VariableExpressionVertex variableExpressionVertex = (VariableExpressionVertex) vertex;
+
+        // try and find the value of this vertex
+        Optional<ApexValue<?>> variableValueOpt =
+                ScopeUtil.resolveToApexValue(symbols, variableExpressionVertex);
+        if (variableValueOpt.isPresent()) {
+            ApexValue<?> variableValue = variableValueOpt.get();
+
+            if (variableValue.hasNegativeConstraint(Constraint.Null)) {
+                // negative constraint means the value is guaranteed NOT to be null,
+                // so this deserves no violation
+                return false;
+            } else if (variableValue.isNull() || variableValue.isIndeterminant()) {
+                // if the variable is null (is actually null or has positive null constraint) OR the
+                // value is indeterminant, this deserves a violation
+                return true;
+            }
+
+        } else {
+            throw new UnexpectedException(
+                    "PerformNullCheckOnSoqlVariables couldn't find an apex value associated with variable vertex "
+                            + vertex);
+        }
+        return false;
+    }
+
+    @Override
+    protected List<RuleThrowable> _run(
+            GraphTraversalSource g, ApexPath path, BaseSFVertex sinkVertex) {
+        if (!(sinkVertex instanceof VariableExpressionVertex)) {
+            throw new ProgrammingException(
+                    "PerformNullCheckOnSoqlVariables rule can only be applied to SoqlExpressionVertex sink vertex. Provided sink vertex="
```

Comment:
Extremely minor nitpick: This message should probably say `VariableExpressionVertex` instead of `SoqlExpressionVertex`, right?